pkg/devicetwin/devicetwin_test.go

```diff
@@ -0,0 +1,148 @@
+	"github.com/kubeedge/kubeedge/pkg/devicetwin/dtcontext"
+	"github.com/kubeedge/kubeedge/pkg/devicetwin/dtmodule"
+	"testing"
+
+	"github.com/golang/mock/gomock"
+
+	"github.com/kubeedge/kubeedge/beehive/pkg/common/log"
+	"github.com/kubeedge/kubeedge/beehive/pkg/core"
+	"github.com/kubeedge/kubeedge/beehive/pkg/core/context"
+	"github.com/kubeedge/kubeedge/beehive/pkg/core/model"
+	"github.com/kubeedge/kubeedge/mocks/beego"
+	"github.com/kubeedge/kubeedge/mocks/beehive"
+	"github.com/kubeedge/kubeedge/pkg/common/dbm"
+)
+
+const (
+	//TestModule is name of test.
+	TestModule = "test"
+	//GroupResource is name of resource group resource.
+	GroupResource = "resource"
+)
+
+// ormerMock is mocked Ormer implementation.
+var ormerMock *beego.MockOrmer
+
+// querySeterMock is mocked QurySeter implementation.
+var querySeterMock *beego.MockQuerySeter
+
+// fakeDeviceTwin is mocked implementation of devicetwin module.
+var fakeDeviceTwin *beehive.MockModule
+
+// mainContext is beehive context used for communication between modules.
+var mainContext *context.Context
+
+// initMocks is function to initialize mocks.
+func initMocks(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+	ormerMock = beego.NewMockOrmer(mockCtrl)
+	querySeterMock = beego.NewMockQuerySeter(mockCtrl)
+	fakeDeviceTwin = beehive.NewMockModule(mockCtrl)
+	dbm.DBAccess = ormerMock
+}
+
+// registerFakeModule is fuction to register all mock modules used.
+func registerFakeModule() {
+	fakeDeviceTwin.EXPECT().Name().Return(TestModule).Times(3)
+	core.Register(fakeDeviceTwin)
+	mainContext = context.GetContext(context.MsgCtxTypeChannel)
+	mainContext.AddModule(TestModule)
+
+}
+
+// TestDeviceTwin_Name is function to test Name().
+func TestDeviceTwin_Name(t *testing.T) {
+	tests := []struct {
+		name string
+		want string
+	}{
+		{
+			name: "edgehub",
+			want: "twin",
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			dt := &DeviceTwin{}
+			if got := dt.Name(); got != tt.want {
+				t.Errorf("DeviceTwin.Name() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+// TestDeviceTwin_Group is function to test Group().
+func TestDeviceTwin_Group(t *testing.T) {
+	tests := []struct {
+		name string
+		want string
+	}{
+		{
+			name: "test1",
+			want: "twin",
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			dt := &DeviceTwin{}
+			if got := dt.Group(); got != tt.want {
+				t.Errorf("DeviceTwin.Group() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+// TestDeviceTwin_Start is function to test Start().
+func TestDeviceTwin_Start(t *testing.T) {
+	initMocks(t)
+	registerFakeModule()
+	core.Register(&DeviceTwin{})
+	dt := DeviceTwin{}
+	mainContext.AddModule(dt.Name())
+	mainContext.AddModuleGroup(dt.Name(), dt.Group())
+	dt.context = mainContext
+	controller, err := InitDTController(mainContext)
+	if err != nil {
+		log.LOGGER.Errorf("Start device twin failed, due to %v", err)
```

Comment:
Can we use t.Errorf instead of log ? no need to continue the test if InitDTController fails